src/main/2.4/scala/com/crealytics/spark/v2/excel/ExcelDataSource.scala

```diff
@@ -0,0 +1,405 @@
+    new ExcelInputPartitionReader(filters, dataSchema, requiredSchema, partitionSchema, options, path, partitionValues)
+}
+
+class ExcelInputPartitionReader(
+  val filters: Array[Filter],
+  val dataSchema: StructType,
+  val requiredSchema: StructType,
+  val partitionSchema: StructType,
+  val options: ExcelOptions,
+  val path: URI,
+  val partitionValues: InternalRow
+) extends InputPartitionReader[InternalRow] {
+
+  /* To populate get_input_filename() */
+  ExcelSparkInternal.setInputFileName(path.getPath)
+
+  private val parser = new ExcelParser(dataSchema, requiredSchema, options, filters)
+  private val headerChecker =
+    new ExcelHeaderChecker(dataSchema, options, source = s"Excel file: ${path}")
+  private val excelHelper = ExcelHelper(options)
+  private val rows = excelHelper.getRows(new Configuration(), path)
+
+  val reader = ExcelParser.parseIterator(rows, parser, headerChecker, dataSchema)
+
+  private val fullSchema = requiredSchema
+    .map(f => AttributeReference(f.name, f.dataType, f.nullable, f.metadata)()) ++
+    partitionSchema.map(f => AttributeReference(f.name, f.dataType, f.nullable, f.metadata)())
+
+  // Using lazy val to avoid serialization
+  private lazy val appendPartitionColumns = GenerateUnsafeProjection
+    .generate(fullSchema, fullSchema)
+
+  private val joinedRow = new JoinedRow()
+
+  private val combinedReader = reader.map(dataRow => {
+    // Using local val to avoid per-row lazy val check (pre-mature optimization?...)
+    val converter = appendPartitionColumns
+    converter(joinedRow(dataRow, partitionValues))
+  })
+
+  override def next: Boolean = combinedReader.hasNext
+  override def get: InternalRow = combinedReader.next
+  override def close(): Unit = {}
+}
+
+class ExcelDataSourceWriter(
+  val sparkSession: SparkSession,
+  val map: Map[String, String],
+  val path: String,
+  val schema: StructType
+) extends DataSourceWriter {
+
+  private lazy val options =
+    new ExcelOptions(map, sparkSession.sessionState.conf.sessionLocalTimeZone)
+
+  override def createWriterFactory(): DataWriterFactory[InternalRow] =
+    new ExcelDataWriterFactory(options, path, schema)
+
+  override def commit(messages: Array[WriterCommitMessage]): Unit = {}
+
+  override def abort(messages: Array[WriterCommitMessage]): Unit = {}
+}
+
+class ExcelDataWriterFactory(val options: ExcelOptions, val path: String, val schema: StructType)
+    extends DataWriterFactory[InternalRow] {
+  override def createDataWriter(partitionId: Int, taskId: Long, epochId: Long): DataWriter[InternalRow] =
+    new ExcelDataWriter(options, (new Path(path, f"part-$partitionId%06d-$taskId%09d.xlsx")).toString(), schema)
+}
+
+class ExcelDataWriter(val options: ExcelOptions, val path: String, val schema: StructType)
+    extends DataWriter[InternalRow] {
+
+  private val gen = new ExcelGenerator(path, schema, new Configuration(), options)
+  if (options.header) { gen.writeHeaders() }
```

Comment:
This means that `ExcelDataWriter` starts writing as soon as it's instantiated, not when the `write` method is called.
Not sure what the contract of a `DataWriter` is, if it can write headers etc. before the first `write` call.

Reply:
Hi @nightscape 
Frankly, I haven't thought about this before. Basically, I was just followed same approach with built-in [CsvOutputWriter.scala](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CsvOutputWriter.scala#L42)
Thinking about this now, I think, when writing with spark data frame, something like this:
```python
df.write.format("excel").save("/an/example/path")
```
spark is lazy by its natural, above paragraph, including how the df is constructed, will be prepared into an execution plan and all this will be actually materialized in the .save(..) action.
At the time of OutputWriter is constructed (on spark executor node) all details about the writing are ready (with header or not, to which path etc). Even if df is empty, writing a header line still makes sense.